src/ints/xms.cpp

```diff
@@ -16,478 +17,843 @@
+{
+	if (!xms_is_handle_valid(handle)) {
+		return Result::XMS_INVALID_HANDLE;
 	}
-	size=(uint16_t)(xms_handles[handle].size);
-	return 0;
+
+	lock_count = xms.handles[handle].lock_count;
+
+	// Find available blocks
+
+	num_free = 0;
+
+	for (uint16_t index = 1; index < NumXmsHandles; ++index) {
+		if (xms.handles[index].is_free) {
+			++num_free;
+		}
+	}
+
+	size_kb = xms.handles[handle].size_kb;
+	return Result::OK;
 }
 
-Bitu XMS_ResizeMemory(Bitu handle, Bitu newSize) {
-	if (InvalidHandle(handle)) return XMS_INVALID_HANDLE;	
+static Result xms_resize_memory(const uint16_t handle, const uint32_t new_size_kb)
+{
+	if (!xms_is_handle_valid(handle)) {
+		return Result::XMS_INVALID_HANDLE;
+	}
+
 	// Block has to be unlocked
-	if (xms_handles[handle].locked>0) return XMS_BLOCK_LOCKED;
-	Bitu pages=newSize/4 + ((newSize & 3) ? 1 : 0);
-	if (MEM_ReAllocatePages(xms_handles[handle].mem,pages,true)) {
-		xms_handles[handle].size = newSize;
-		return 0;
-	} else return XMS_OUT_OF_SPACE;
+
+	if (xms.handles[handle].lock_count > 0) {
+		return Result::XMS_BLOCK_LOCKED;
+	}
+
+	constexpr bool sequence = true;
+	if (MEM_ReAllocatePages(xms.handles[handle].mem_handle,
+	                        get_num_pages(new_size_kb),
+	                        sequence)) {
+		xms.handles[handle].size_kb = new_size_kb;
+		return Result::OK;
+	}
+
+	return Result::XMS_OUT_OF_SPACE;
 }
 
-static bool multiplex_xms(void) {
+static bool xms_multiplex()
+{
 	switch (reg_ax) {
-	case 0x4300:					/* XMS installed check */
-			reg_al=0x80;
-			return true;
-	case 0x4310:					/* XMS handler seg:offset */
-			SegSet16(es,RealSeg(xms_callback));
-			reg_bx=RealOff(xms_callback);
-			return true;			
+	case 0x4300: // XMS installed check
+		reg_al = 0x80;
+		return true;
+	case 0x4310: // XMS handler seg:offset
+		SegSet16(es, RealSeg(xms.callback));
+		reg_bx = RealOff(xms.callback);
+		return true;
 	}
-	return false;
 
+	return false;
 }
 
-inline void SET_RESULT(Bitu res,bool touch_bl_on_succes=true) {
-	if(touch_bl_on_succes || res) reg_bl = (uint8_t)res;
-	reg_ax = (res==0);
-}
+// ***************************************************************************
+// Main XMS API handler
+// ***************************************************************************
+
+static Bitu XMS_Handler()
+{
+	assert(xms.is_available);
+
+	Result result = Result::OK;
+
+	auto set_return_value = [](const Result result) {
+		reg_bl = static_cast<uint8_t>(result);
+		reg_ax = (result == Result::OK) ? 1 : 0;
+	};
+
+	auto set_return_value_bl_only_fail = [](const Result result) {
+		if (result != Result::OK) {
+			reg_bl = static_cast<uint8_t>(result);
+		}
+		reg_ax = (result == Result::OK) ? 1 : 0;
+	};
 
-Bitu XMS_Handler(void) {
-//	LOG(LOG_MISC,LOG_ERROR)("XMS: CALL %02X",reg_ah);
 	switch (reg_ah) {
-	case XMS_GET_VERSION:										/* 00 */
-		reg_ax=XMS_VERSION;
-		reg_bx=XMS_DRIVER_VERSION;
-		reg_dx=0;	/* No we don't have HMA */
+	case 0x00: // Get XMS Version Number
```

Comment:
Several new functions implemented, and code is much more readable now :) Great job!